File: Linkedlist/reverse_linkedlist.kt
package LinkedList


// O(n)
internal class Solution {
    fun reverseList(head: ListNode?): ListNode? {
        // Null for the head
        var prev: ListNode? = null
        var curr: ListNode? = head
        while (curr != null) {

            // Use a temp to hold the next variable
            val nextTemp: ListNode? = curr.next
            curr.next = prev
            prev = curr
            curr = nextTemp
        }
        return prev
    }
}

File: Linkedlist/reverseLinkedlist.py

# This is the leet code accepted answer
# Reverse a singly linkedlist, need 3 pointer here
def reverseSinglyLinkedList(head):

    if not head:
        return None

    cur = head
    prev = None

    while cur:
        next = cur.next
        cur.next = prev
        prev = cur
        cur = next
    return prev

